File: week-03_05/tp/tp_C/tp7/solution/creation/creation.c
#include <mpi.h>
#include <stdio.h>

int main(int argc, char *argv[]) {
  const int nb_valeurs=242;
  int valeurs[2*nb_valeurs];
  FILE *unit;
  int iter;
  MPI_File descripteur;
  MPI_Status statut;

  MPI_Init(&argc, &argv);

  unit = fopen("carte.dat","r");
  /* Read X Y positions */
  for (iter=0; iter<nb_valeurs; iter++) {
    if (fscanf(unit,"%d %d\n",
               &(valeurs[iter]), &(valeurs[nb_valeurs+iter])) != 2) {
      fprintf(stderr,"Error reading file\n");
      MPI_Abort(MPI_COMM_WORLD,1); }
    };
  fclose(unit);
  /* Values=All X positions then All Y positions */
  MPI_File_set_errhandler(MPI_FILE_NULL,MPI_ERRORS_ARE_FATAL);
  MPI_File_open(MPI_COMM_WORLD,"data.dat",MPI_MODE_RDWR | MPI_MODE_CREATE,
                MPI_INFO_NULL,&descripteur);
  MPI_File_write_all(descripteur,valeurs,2*nb_valeurs,MPI_INT,&statut);
  MPI_File_close(&descripteur);
  MPI_Finalize();
  return 0;
}

File: week-03_05/tp/tp/tp7/creation/creation.f90
! On 1 process !

program creation

  use mpi_f08

  implicit none

  integer, parameter              :: nb_values=242
  integer, dimension(2*nb_values) :: values
  type(MPI_Status)                :: status
  integer                         :: i,rank
  type(MPI_File)                  :: fh

  open(unit=45,file="carte.dat")
  ! Read X Y positions
  do i=1,nb_values
    read(45,*) values(i),values(nb_values+i)
  end do
  close(45)
  ! Values=All X positions then All Y positions

  call MPI_INIT()
  call MPI_COMM_RANK(MPI_COMM_WORLD,rank)
  call MPI_FILE_SET_ERRHANDLER(MPI_FILE_NULL,MPI_ERRORS_ARE_FATAL)
  call MPI_FILE_OPEN(MPI_COMM_WORLD,"data.dat",MPI_MODE_RDWR + MPI_MODE_CREATE, &
                     MPI_INFO_NULL,fh)

  call MPI_FILE_WRITE_ALL(fh,values,2*nb_values,MPI_INTEGER, &
                          status)

  call MPI_FILE_CLOSE(fh)
  call MPI_FINALIZE()

end program creation
